### ISSUE TITLE
Tier 1 support for `x86_64-linux-(gnu,musl)x32`

### ISSUE THREAD

**Issue by alexrp** [ziglang/zig#21711]
- [ ] Standard library support
- [ ] libc is available even when cross-compiling
- [ ] Stack traces work
- [ ] All behavior tests pass
- [ ] All standard library tests pass
- [ ] No disabled tests
- [ ] Enabled in CI (QEMU)

Open issues:

* https://github.com/ziglang/zig-bootstrap/issues/20

**Comment by nektro** [comment]
found this https://lwn.net/Articles/774734/ the last time i looked into it, which if true may make it tier 4 by default

**Comment by alexrp** [comment]
`git log --grep=x32 arch/x86` in the Linux tree suggests it still sees ongoing development. If that changes, though, I have no particular objection to downgrading its support status.

**Comment by rootbeer** [comment]
I've been looking into Zig standard library support for X32, and one issue I've run into is that `usize` is 32-bit, but X32 syscalls take 64-bit parameters (see https://man7.org/linux/man-pages/man2/syscall.2.html).   Changing the x86_64.zig `syscall*` wrappers is straight-forward, but the deeper issue becomes the common `lib/std/os/linux.zig` code that maps Zig APIs into syscall invocations.  Currently all that code (and probably some code in File.zig and posix.zig, too), assumes syscall parameters are `usize`.  For example, the `lseek`  wrapper does ` @as(usize, @bitCast(@as(isize, fd)))` to sign-extend a `fd` into the syscall register.  But on X32 `usize` is 32-bit, so this isn't right.

So, I think adding X32 support will require changes throughout `linux.zig`, as the common code can't assume syscall registers are `usize` anymore.

Do we have examples of other platforms in Zig that handle syscalls like this?  Or do you know of examples in other systems that we should follow the lead on?

My inclination is to add a `syscall_param` (`SyscallParam`?) type that is `usize` on most Linux platforms, but is `u64` on X32.  And maybe add a generic `syscall_encode` utility in linux.zig that takes signed/unsigned integers and does the right bitcasts into the platform's syscall parameter type.

If you don't yet have any opinions on this, I'll put together a change that demonstrates the basics before I touch all of `linux.zig`.  But before that, there are a couple other fixes I should send out.

**Comment by alexrp** [comment]
> Do we have examples of other platforms in Zig that handle syscalls like this?

No, this is a first. But it seems likely that MIPS N32 will be similar: #21712

> My inclination is to add a `syscall_param` (`SyscallParam`?) type that is `usize` on most Linux platforms, but is `u64` on X32. And maybe add a generic `syscall_encode` utility in linux.zig that takes signed/unsigned integers and does the right bitcasts into the platform's syscall parameter type.

This sounds like a sensible direction to me.
